platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentGraphResolveState.java

```diff
@@ -85,30 +87,42 @@ public DefaultLocalComponentGraphResolveState(
         ComponentIdGenerator idGenerator,
         boolean adHoc,
         ConfigurationMetadataFactory configurationFactory,
+        CalculatedValueContainerFactory calculatedValueContainerFactory,
         @Nullable Transformer<LocalComponentArtifactMetadata, LocalComponentArtifactMetadata> artifactTransformer
     ) {
         super(instanceId, metadata, attributeDesugaring);
         this.idGenerator = idGenerator;
         this.adHoc = adHoc;
         this.configurationFactory = configurationFactory;
+        this.calculatedValueContainerFactory = calculatedValueContainerFactory;
         this.artifactTransformer = artifactTransformer;
 
-        // TODO: We should be using the CalculatedValue infrastructure to lazily compute these values
-        //       in order to properly manage project locks.
-        this.graphSelectionCandidates = Lazy.locking().of(() ->
-            computeGraphSelectionCandidates(this, idGenerator, configurationFactory, artifactTransformer)
-        );
-        this.selectableVariantResults = Lazy.locking().of(() ->
-            computeSelectableVariantResults(this)
-        );
+        // Mutable state
+        this.configurations = calculatedValueContainerFactory.createCache(Describables.of("configurations"));
+        initCalculatedValues();
     }
 
     @Override
     public void reevaluate() {
-        // TODO: This is not thread-safe. We do not atomically clear all the different fields at once.
+        // TODO: This is not really thread-safe.
+        //       We should atomically clear all the different fields at once.
+        //       Or better yet, we should not allow reevaluation of the state.
         configurations.clear();
         configurationFactory.invalidate();
-        // TODO: We are missing logic to invalidate allVariantsForGraphResolution and selectableVariantResults.
+        initCalculatedValues();
+    }
+
+    private void initCalculatedValues() {
+        this.graphSelectionCandidates.set(
```

Comment:
Needing to wrap these calculated values in `AtomicReference` is a terrible solution, but as it stands the prior implementation did not invalidate the list of variants on when `reevaluate` was called. 

Indeed, the proper solution is to get rid of `reevaluate`, and this has [already been addressed](https://github.com/gradle/gradle/pull/27558) by deprecating the cases that cause us to need to reevaluate. 

For now, however, the prior behavior of not invalidating this state is the cause of other issues, particularly [this one from Spring](https://github.com/gradle/gradle/issues/27947) where their dependency management plugin relies on mutating configurations in a `beforeResolve` hook. In these cases, the resolution engine can see invalid definitions of a variant when the configuration was first resolved for build dependencies and then is resolved later for artifacts.